run/logging-manual/e2e_test.py

```diff
@@ -0,0 +1,156 @@
+
+# This sample creates a secure two-service application running on Cloud Run.
+# This test builds and deploys the two secure services
+# to test that they interact properly together.
+
+import json
+import os
+import subprocess
+import time
+from urllib import request
+import uuid
+
+import pytest
+
+
+@pytest.fixture()
+def services():
+    # Unique suffix to create distinct service names
+    suffix = uuid.uuid4().hex
+    project = os.environ["GOOGLE_CLOUD_PROJECT"]
+
+    # Build and Deploy Cloud Run Services
+    subprocess.run(
+        [
+            "gcloud",
+            "builds",
+            "submit",
+            "--project",
+            project,
+            "--substitutions",
+            f"_SUFFIX={suffix}",
+            "--config",
+            "e2e_test_setup.yaml",
+            "--quiet",
+        ],
+        check=True,
+    )
+
+    # Get the URL for the service and the token
+    service = subprocess.run(
+        [
+            "gcloud",
+            "run",
+            "--project",
+            project,
+            "--platform=managed",
+            "--region=us-central1",
+            "services",
+            "describe",
+            f"logging-{suffix}",
+            "--format=value(status.url)",
+        ],
+        stdout=subprocess.PIPE,
+        check=True,
+    ).stdout.strip()
+
+    id_token = subprocess.run(
+        ["gcloud", "auth", "print-identity-token"], stdout=subprocess.PIPE, check=True
+    ).stdout.strip()
+
+    access_token = subprocess.run(
+        ["gcloud", "auth", "print-access-token"], stdout=subprocess.PIPE, check=True
+    ).stdout.strip()
+
+    yield service, id_token, access_token, project, f"logging-{suffix}"
+
+    subprocess.run(
+        [
+            "gcloud",
+            "run",
+            "services",
+            "delete",
+            f"logging-{suffix}",
+            "--project",
+            project,
+            "--platform",
+            "managed",
+            "--region",
+            "us-central1",
+            "--quiet",
+        ],
+        check=True,
+    )
+
+
+def test_end_to_end(services):
+    service = services[0].decode()
+    id_token = services[1].decode()
+    access_token = services[2].decode()
+    project = services[3]
+    service_name = services[4]
+
+    # Test that the service is responding
+    req = request.Request(
+        service,
+        headers={
+            "Authorization": f"Bearer {id_token}",
+            "X-Cloud-Trace-Context": "foo/bar",
+        },
+    )
+    response = request.urlopen(req)
+    assert response.status == 200
+
+    body = response.read()
+    assert body.decode() == "Hello Logger!"
+
+    # Test that the logs are writing properly to stackdriver
+    time.sleep(10)  # Slight delay writing to stackdriver
+    response = stackdriver_request(service_name, project, access_token)
+
+    entries = json.loads(response).get("entries")
+    assert entries
+
+
+def stackdriver_request(service_name, project, access_token):
```

Comment:
1. Why not use client library?
2. I notice no retries, it's been found necessary in Node.js and Golang to avoid flakiness:
  * https://github.com/GoogleCloudPlatform/nodejs-docs-samples/blob/master/run/logging-manual/test/system.test.js#L20
  * https://github.com/GoogleCloudPlatform/golang-samples/blob/master/logging/simplelog/simplelog_test.go#L63

Reply:
Updated to add retries.

Updated to the client library.  The reason why I didn't use it as first is because  doing the direct HTTP calls is generally easier. The documentation for the client libraries is often not great.  By contrast, I can use the API explorer to figure out the correct API call and then copy the HTTP request from the UI.